### ISSUE TITLE
Allow the passing of arrays with both functions and strings to the permission system 

### ISSUE THREAD

**Issue by Discordius** [VulcanJS/Vulcan#2002]
**Copied from Slack:** 
Feature request for the permission system: Allow `viewableBy`, `editableBy` and `insertableBy` to take arrays of functions as well as arrays of group names
I have a pretty common use-case where I want certain fields to be accessible by the owner of the document, and by moderators
But right now there is no elegant way to make that happen
I can either make it accessible by moderators, by passing `['moderators']`
Or I can make it accessible by the owners, by passing `Users.owns`
But I can't do both, i.e. this doesn't work: `[Users.owns, 'moderators']`

**Comment by Apollinaire** [comment]
Did the code for permissions change since the `opencrud` update ? If it did not, I could work on this, since it is useful to me too (I have the same use case as @Discordius) but I have not updated yet and won't before a couple of months

**Comment by SachaG** [comment]
It did change a bit so I'd recommend working on the latest `devel` branch if you want to submit a PR for that.

**Comment by Apollinaire** [comment]
Alrigth, I'll work on my older version first (1.9) and will do it after with the most recent version

**Comment by Apollinaire** [comment]
Actually I don't see what changed on the way we treat `viewableBy, editableBy, insertableBy`, to me the functions [`Users.canViewField`](https://github.com/VulcanJS/Vulcan/blob/a2e3ca01cc6de29a6138aee429ef6c9598c8a023/packages/vulcan-users/lib/modules/permissions.js#L192), [`Users.canEditFields`](https://github.com/VulcanJS/Vulcan/blob/a2e3ca01cc6de29a6138aee429ef6c9598c8a023/packages/vulcan-users/lib/modules/permissions.js#L268), [`Users.canInsertFields`](https://github.com/VulcanJS/Vulcan/blob/a2e3ca01cc6de29a6138aee429ef6c9598c8a023/packages/vulcan-users/lib/modules/permissions.js#L256) did not change ? Am I missing something? 

If I'm not missing anything, i have made this that seems to work with v1.9.1 : 
```js
Users.canViewField = function (user, field, document) {
  if (field.viewableBy) {
    if (typeof field.viewableBy === 'function') {
      // if viewableBy is a function, execute it with user and document passed. it must return a boolean
      return field.viewableBy(user, document);
    } else if (typeof field.viewableBy === 'string') {
      // if viewableBy is just a string, we assume it's the name of a group and pass it to isMemberOf
      return Users.isMemberOf(user, field.viewableBy);
    } else if (Array.isArray(field.viewableBy) && field.viewableBy.length > 0) {
      // if viewableBy is an array, we do a recursion on every item and return true if one of the items return true
      // this also makes it possible to use nested arrays, such as ['admins', ['group1', function1, [function2, 'group2'], function3]]
      return field.viewableBy.reduce((accumulator, currentValue)=> accumulator || Users.canViewField(user, currentValue, document));
    }
  }
  return false;
};
```

It uses recursion to treat arrays. Any feedback is welcome! 
The functions for edit and insert will be the same

Just FYI, the current function is: 
```js
Users.canViewField = function (user, field, document) {
  if (field.viewableBy) {
    return typeof field.viewableBy === 'function' ? field.viewableBy(user, document) : Users.isMemberOf(user, field.viewableBy)
  }
  return false;
};
```

**Comment by Apollinaire** [comment]
Ok, so in order to handle the new API I must also take care of a possible `field.canRead` in my function above, right ? 
( In the code I did not find anything else than the [simple-schema extension config](https://github.com/VulcanJS/Vulcan/blob/973b2bfce34065d9fd2f740df8592dc6da0ff665/packages/vulcan-lib/lib/modules/config.js#L35) about these, are they handled already or is it still a WIP?)

**Comment by SachaG** [comment]
You're right it's still a WIP. For some reason I thought (hoped?) I had already done it…

**Comment by SachaG** [comment]
And yeah I think you can just do something like

```
const canRead = field.canRead || field.viewableBy;
if (canRead) ...
```

Just make sure you leave this comment whenever you're adding backward compatibility: 

```
// OpenCRUD backwards compatibility
```

**Comment by Apollinaire** [comment]
Sure, will do. PR coming soon

**Comment by Apollinaire** [comment]
Actually, while i'm working on this code, maybe update the name of the functions according to the new OpenCRUD names? 
Here's what I think might be the best: 

Field name | Function name
---------------- | ------------
viewableBy | Users.canViewField
canRead | Users.canReadField
insertableBy | Users.canInsertField
canCreate | Users.canCreateField
editableBy | Users.canEditField
canUpdate | Users.canUpdateField

obviously we keep backwards compatibility for the old names

**Comment by SachaG** [comment]
Correct me if I'm wrong but I think those functions would mostly be used by Vulcan itself internally? I don't think they're documented? If so we don't need to preserve backwards compatibility I think.

**Comment by Apollinaire** [comment]
You're right, I don't think they're documented. Still, they're easily available to anyone who would look in the code and reuse it in their own code.
If we don't keep it backwards compatible, we can just say we changed the name in the release notes to warn anyone who might be using this.

**Comment by SachaG** [comment]
Yes I think for those a warning in the release notes is sufficient.

**Comment by kevinashworth** [comment]
I've been seeing unexpected behavior on my system, and when I use this code in permissions.js, it seems to work:

` Users.canReadField = function ( user, field, document) {
   const canRead = field.canRead || field.viewableBy; //OpenCRUD backwards compatibility
   if (canRead) {
     if (typeof canRead === 'function') {
       // if canRead is a function, execute it with user and document passed. it must return a boolean
       return canRead(user, document);
     } else if (typeof canRead === 'string') {
       if (canRead == 'guests') {
         return true;
       }
       // if canRead is just a string, we assume it's the name of a group and pass it to isMemberOf
       return Users.isMemberOf(user, canRead);
     } else if (Array.isArray(canRead) && canRead.length > 0) {
       // if canRead is an array, we do a recursion on every item and return true if one of the items return true
       // this also makes it possible to use nested arrays, such as ['admins', ['group1', function1, [function2, 'group2'], function3]]
       const reducer = (accumulator, currentValue) => accumulator || Users.canReadField(user, {canRead: currentValue}, document);
       return canRead.reduce(reducer, false);
     }
   }
   return false;
 };`

The changes: 

1 String comparison to 'guests' means fields marked `'guests'`, `['guests']`, `['guests', 'members']` and `['members']` can now all be read by members.

2 The accumulator was returning a string so arrays aren't being handled correctly. I added an initialValue of `false` and made the middle argument of the recursion call to be an object with key `canRead`.
